nav2_map_server/src/occ_grid_loader.cpp

```diff
@@ -27,100 +27,195 @@
+    loadParameters.mode = TRINARY;
   }
 
   try {
-    negate_ = doc_["negate"].as<int>();
+    loadParameters.negate = doc["negate"].as<int>();
   } catch (YAML::Exception) {
     throw std::runtime_error("The map does not contain a negate tag or it is invalid");
   }
 
-  RCLCPP_DEBUG(node_->get_logger(), "resolution: %f", resolution_);
-  RCLCPP_DEBUG(node_->get_logger(), "origin[0]: %f", origin_[0]);
-  RCLCPP_DEBUG(node_->get_logger(), "origin[1]: %f", origin_[1]);
-  RCLCPP_DEBUG(node_->get_logger(), "origin[2]: %f", origin_[2]);
-  RCLCPP_DEBUG(node_->get_logger(), "free_thresh: %f", free_thresh_);
-  RCLCPP_DEBUG(node_->get_logger(), "occupied_thresh: %f", occupied_thresh_);
+  RCLCPP_DEBUG(node_->get_logger(), "resolution: %f", loadParameters.resolution);
+  RCLCPP_DEBUG(node_->get_logger(), "origin[0]: %f", loadParameters.origin[0]);
+  RCLCPP_DEBUG(node_->get_logger(), "origin[1]: %f", loadParameters.origin[1]);
+  RCLCPP_DEBUG(node_->get_logger(), "origin[2]: %f", loadParameters.origin[2]);
+  RCLCPP_DEBUG(node_->get_logger(), "free_thresh: %f", loadParameters.free_thresh);
+  RCLCPP_DEBUG(node_->get_logger(), "occupied_thresh: %f", loadParameters.occupied_thresh);
   RCLCPP_DEBUG(node_->get_logger(), "mode_str: %s", mode_str.c_str());
-  RCLCPP_DEBUG(node_->get_logger(), "mode: %d", mode_);
-  RCLCPP_DEBUG(node_->get_logger(), "negate: %d", negate_);
+  RCLCPP_DEBUG(node_->get_logger(), "mode: %d", loadParameters.mode);
+  RCLCPP_DEBUG(node_->get_logger(), "negate: %d", loadParameters.negate);
+
+  loadMapFromFile(map_filename, &loadParameters);
+
+  // Create a service callback handle
+  auto handle_occ_callback = [this](
+    const std::shared_ptr<rmw_request_id_t>/*request_header*/,
+    const std::shared_ptr<nav_msgs::srv::GetMap::Request>/*request*/,
+    std::shared_ptr<nav_msgs::srv::GetMap::Response> response) -> void {
+      RCLCPP_INFO(node_->get_logger(), "OccGridLoader: Handling map request");
+      response->map = *msg_;
+    };
+
+  // Create a service that provides the occupancy grid
+  occ_service_ = node_->create_service<nav_msgs::srv::GetMap>(service_name_, handle_occ_callback);
+
+  // Create a publisher using the QoS settings to emulate a ROS1 latched topic
+  rmw_qos_profile_t custom_qos_profile = rmw_qos_profile_default;
+  custom_qos_profile.depth = 1;
+  custom_qos_profile.durability = RMW_QOS_POLICY_DURABILITY_TRANSIENT_LOCAL;
+  custom_qos_profile.reliability = RMW_QOS_POLICY_RELIABILITY_RELIABLE;
+  occ_pub_ = node_->create_publisher<nav_msgs::msg::OccupancyGrid>(
+    topic_name_, custom_qos_profile);
+
+  return nav2_lifecycle::CallbackReturn::SUCCESS;
+}
+
+nav2_lifecycle::CallbackReturn
+OccGridLoader::on_activate(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(node_->get_logger(), "OccGridLoader: Activating");
+
+  // Publish the map using the latched topic
+  occ_pub_->on_activate();
+  occ_pub_->publish(*msg_);
```

Comment:
Before we had a timer and were publishing the map periodically, do we not need to do that anymore?

Reply:
If I remember correctly, I believe we were publishing the map periodically before in order to interface across the ROS1 bridge since latched topics didn't work the same way.

Reply:
Right, I think we still need to support publishing over the ROS1 bridge. Maybe we add a parameter to map_server called 'loop_publisher" or something and we loop if it's set to true.

Reply:
We discussed and I agree we can remove this. I don't know of a usage that requires running map_server across the bridge. If we come across one, we can add this back in. I'm marking this resolved.